marlowe-test/src/Spec/Marlowe/Plutus/MList.hs

```diff
@@ -0,0 +1,305 @@
+--   "lookup a Nil = None" |
+--   "lookup a (Cons (x, y) z) =
+--     (if a = x
+--      then Some y
+--      else (if a > x
+--            then lookup a z
+--            else None))"
+-- @
+lookup :: Ord a => a -> MList a b -> Maybe b
+lookup _a [] = Nothing
+lookup a ((x, y) : z) =
+  if a == x
+    then Just y
+    else if a > x
+      then lookup a z
+      else Nothing
+
+
+{-# ANN unionWith "HLint: ignore Use guards" #-}
+
+-- | Union with Isabelle `MList`.
+--
+-- @
+-- fun unionWith :: "('b \<Rightarrow> 'b \<Rightarrow> 'b) \<Rightarrow> ('a \<times> 'b) list \<Rightarrow>
+--                   ('a \<times> 'b) list \<Rightarrow> (('a::linorder) \<times> 'b) list" where
+--   "unionWith f (Cons (x, y) t) (Cons (x2, y2) t2) =
+--     (if x < x2
+--      then Cons (x, y) (unionWith f t (Cons (x2, y2) t2))
+--      else (if x > x2
+--            then Cons (x2, y2) (unionWith f (Cons (x, y) t) t2)
+--            else Cons (x, f y y2) (unionWith f t t2)))" |
+--   "unionWith f Nil l = l" |
+--   "unionWith f l Nil = l"
+-- @
+unionWith :: Ord a => (b -> b -> b) -> MList a b -> MList a b -> MList a b
+unionWith f ((x, y) : t) ((x2, y2) : t2) =
+  if x < x2
+    then (x, y) : unionWith f t ((x2, y2) : t2)
+    else if x > x2
+      then (x2, y2) : unionWith f ((x, y) : t) t2
+      else (x, f y y2) : unionWith f t t2
+unionWith _f [] l = l
+unionWith _f l [] = l
+
+
+-- | Find with default for Isabelle `MList`.
+--
+-- @
+-- fun findWithDefault :: "'b \<Rightarrow> 'a \<Rightarrow> (('a::linorder) \<times> 'b) list \<Rightarrow> 'b" where
+--   "findWithDefault d k l = (case lookup k l of
+--                               None \<Rightarrow> d
+--                             | Some x \<Rightarrow> x)"
+-- @
+findWithDefault :: Ord a => b -> a -> MList a b -> b
+findWithDefault d k l =
+  case lookup k l of
+    Nothing -> d
+    Just x -> x
+
+
+-- | Run tests.
+tests :: TestTree
+tests =
+  testGroup "MList vs AssocMap"
+    [
+      testCase     "`empty` is equivalent"           checkEmpty
+    , testProperty "`null` is equivalent"            checkNull
+    , testProperty "`singleton` is equivalent"       checkSingleton
+    , testProperty "`insert` is equivalent"          checkInsert
+    , testProperty "`delete` is equivalent"          checkDelete
+    , testProperty "`lookup` is equivalent"          checkLookup
+    , testProperty "`member` is equivalent"          checkMember
+    , testProperty "`unionWith` is equivalent"       checkUnionWith
+    , testProperty "`findWithDefault` is equivalent" checkFindWithDefault
+    ]
+
+
+-- | Generate a sorted `MList` with no duplicates.
+arbitraryMList :: Gen (MList Integer [()])
+arbitraryMList = nubBy ((==) `on` fst) . sortBy (compare `on` fst) <$> arbitrary
+
+
+-- | Compare an `MList` to an `AssocMap`.
+equivalent :: Ord a => Eq b => MList a b -> AM.Map a b -> Bool
+equivalent mlist assocmap = mlist == sortBy (compare `on` fst) (AM.toList assocmap)
+
```

Comment:
instead of comparing by equality of the sorted AssocMap, shouldn't we check that for each entry in MList Assoc.lookup gives a result and vice-versa?

This way is not checking for duplicate entries in AssocMap (see my comment on insert)

Reply:
Fixed in b65ccfec3708484ff52514671e0e227c13f084f3 by including `lookup` tests, too.

Note that the testing is preconditioned on the `MList` being sorted and neither list containing duplicates. I'll noted that in a revised comment.